fix!: close epochs based on block numbers (#445)
cartesi/rollups-node: CHANGELOG.md

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

### Changed

- Changed the dispatcher to close epochs based on block numbers instead of block timestamps.

> - Renamed `CARTESI_EPOCH_DURATION` to `CARTESI_EPOCH_LENGTH`, and set its default value to 5760 (1 day worth of blocks, in average).

Reviewer: Please use `CARTESI_EPOCH_BLOCK_LENGTH` instead.

```suggestion
- **BREAKING**: replaced `CARTESI_EPOCH_DURATION` with `CARTESI_EPOCH_BLOCK_LENGTH` to match the new epoch calculation procedure, and set its default value to 5760 (1 day worth of blocks, in average).
```

Reviewer 2: I also feel there should be a reference to the `ClaimMismatch` error being fixed. Something like:

```
### Fixed
- Fixed a bug that caused the node to create a ClaimMismatch error when reprocessing inputs after a reboot.
```

Reviewer 2: I also feel there should be a reference to the `ClaimMismatch` bug being fixed. Something like:

```
### Fixed
- Fixed a bug that caused a `ClaimMismatch` error when reprocessing inputs after a reboot.
```

- Added Rollups end-to-end test using Echo Dapp

### Fixed

- Fixed a bug that caused a `authority_claimer::ClaimMismatch` error when reprocessing inputs after a reboot.

### Changed

- Changed the dispatcher to close epochs based on block numbers instead of block timestamps.
- **BREAKING**: replaced the environment variable `CARTESI_EPOCH_DURATION` with `CARTESI_EPOCH_LENGTH_IN_BLOCKS` to match the new epoch algorithm, and set its default value to 7200 (1 day worth of blocks, in average, considering one block is mined every 12 seconds).

> - **BREAKING**: replaced the internal environment variable `RD_EPOCH_DURATION` with `RD_EPOCH_LENGTH_IN_BLOCKS` to match the new epoch algorithm, and also set its default value to 7200.

Reviewer: Being an internal configuration, this would not require a changelog entry. I believe we overdid it here.

- Added Rollups end-to-end test using Echo Dapp

### Fixed

- Fixed a bug that caused a `authority_claimer::ClaimMismatch` error when reprocessing inputs after a reboot.

### Changed

- Changed the dispatcher to close epochs based on block numbers instead of block timestamps.

> - **BREAKING**: replaced the environment variable `CARTESI_EPOCH_DURATION` with `CARTESI_EPOCH_LENGTH_IN_BLOCKS` to match the new epoch algorithm, and set its default value to 7200 (1 day worth of blocks, in average, considering one block is mined every 12 seconds).

Reviewer: ```suggestion
- **BREAKING**: replaced the environment variable `CARTESI_EPOCH_DURATION` with `CARTESI_EPOCH_LENGTH_IN_BLOCKS` to match the new epoch algorithm, and set its default value to 7200 (1 day worth of blocks, considering one block is mined every 12 seconds).
```

- Added Rollups end-to-end test using Echo Dapp

### Fixed

- Fixed a bug that caused a `authority_claimer::ClaimMismatch` error when reprocessing inputs after a reboot.

### Changed

- Changed the dispatcher to close epochs based on block numbers instead of block timestamps.
- **BREAKING**: changed the dispatcher to use the `CARTESI_CONTRACTS_INPUT_BOX_DEPLOYMENT_BLOCK_NUMBER` environment variable instead of `CARTESI_CONTRACTS_APPLICATION_DEPLOYMENT_BLOCK_NUMBER`.

> - **BREAKING**: replaced the environment variable `CARTESI_EPOCH_DURATION` with `CARTESI_EPOCH_LENGTH` to match the new epoch algorithm, and set its default value to 7200 (1 day worth of blocks, in average, considering one block is mined every 12 seconds).

Reviewer: ```suggestion
- **BREAKING**: replaced the environment variable `CARTESI_EPOCH_DURATION` with `CARTESI_EPOCH_LENGTH` to match the new epoch algorithm, and set its default value to 7200 (1 day worth of blocks, considering one block is mined every 12 seconds).
```
